src/charm.py

```diff
@@ -3,113 +3,97 @@
+        try:
+            self.jenkins_agent_service.install()
+        except service.PackageInstallError as e:
+            logger.debug("Error installing the agent service %s", e)
+            self.unit.status = ops.ErrorStatus("Error installing the agent service")
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
+        """Handle config changed event. Update the agent's label in the relation's databag."""
+        if agent_relation := self.model.get_relation(AGENT_RELATION):
+            relation_data = self.state.agent_meta.get_jenkins_agent_v0_interface_dict()
+            agent_relation.data[self.unit].update(relation_data)
+
+    def _on_upgrade_charm(self, event: ops.UpgradeCharmEvent) -> None:
+        """Handle upgrade charm event.
 
         Args:
-            event: event triggering the handler.
+            event: The event fired on upgrade charm.
         """
-        # Fetch the new config value
-        log_level = self.model.config["log-level"].lower()
-
-        # Do some validation of the configuration option
-        if log_level in VALID_LOG_LEVELS:
-            # The config is good, so update the configuration of the workload
-            container = self.unit.get_container("httpbin")
-            # Verify that we can connect to the Pebble API in the workload container
-            if container.can_connect():
-                # Push an updated layer with the new config
-                container.add_layer("httpbin", self._pebble_layer, combine=True)
-                container.replan()
-
-                logger.debug("Log level for gunicorn changed to '%s'", log_level)
-                self.unit.status = ops.ActiveStatus()
-            else:
-                # We were unable to connect to the Pebble API, so we defer this event
-                event.defer()
-                self.unit.status = ops.WaitingStatus("waiting for Pebble API")
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = ops.BlockedStatus("invalid log level: '{log_level}'")
-
-    @property
-    def _pebble_layer(self):
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
-            "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {
-                        "GUNICORN_CMD_ARGS": f"--log-level {self.model.config['log-level']}"
-                    },
-                }
-            },
-        }
-
-
-if __name__ == "__main__":  # pragma: nocover
-    ops.main.main(IsCharmsTemplateCharm)
+        self.reconcile(event)
+
+    def reconcile(self, _: ops.EventBase) -> None:
```

Comment:
Perhaps name this differently for what it is doing, e.g., starting the agent? I'm not sure how reconcile is relevant since that is more of a concept of matching the requested state to the actual state?

Reply:
I renamed this method to a more relevant name (_restart)